Lib/pdb.py

```
@@ -2491,6 +2507,582 @@ def set_trace(*, header=None, commands=None):
+
+    if Pdb._last_pdb_instance is not None:
+        remote_pdb.error("Another PDB instance is already attached.")
+    elif version != remote_pdb.protocol_version():
+        target_ver = f"0x{remote_pdb.protocol_version():08X}"
+        attach_ver = f"0x{version:08X}"
+        remote_pdb.error(
+            f"The target process is running a Python version that is"
+            f" incompatible with this PDB module."
+            f"\nTarget process pdb protocol version: {target_ver}"
+            f"\nLocal pdb module's protocol version: {attach_ver}"
+        )
+    else:
+        remote_pdb.rcLines.extend(commands.splitlines())
+        remote_pdb.set_trace(frame=frame)
+
+
+def attach(pid, commands=()):
+    """Attach to a running process with the given PID."""
+    with closing(socket.create_server(("localhost", 0))) as server:
+        port = server.getsockname()[1]
+
+        with tempfile.NamedTemporaryFile("w", delete_on_close=False) as connect_script:
+            connect_script.write(
+                textwrap.dedent(
+                    f"""
+                    import pdb, sys
+                    pdb._connect(
+                        host="localhost",
+                        port={port},
+                        frame=sys._getframe(1),
+                        commands={json.dumps("\n".join(commands))},
+                        version={_PdbServer.protocol_version()},
+                    )
+                    """
+                )
+            )
+            connect_script.close()
+            sys.remote_exec(pid, connect_script.name)
+
+            # TODO Add a timeout? Or don't bother since the user can ^C?
+            client_sock, _ = server.accept()
+
+            with closing(client_sock):
+                sockfile = client_sock.makefile("rwb")
+
+            with closing(sockfile):
+                with tempfile.NamedTemporaryFile("w", delete_on_close=False) as interrupt_script:
+                    interrupt_script.write(
```

Okay I think I found a problem here. If you do

```python
while True:
    pass
```

from client, the server will stuck in the loop forever.

Because `set_trace()` will not break in `pdb` - it's already tracing.

I thought you could simply do a `signal.raise_signal(signal.SIGINT)` here, but it seems like the exceptions inside `sys.exec_remote` is not propagated?

---

Exceptions aren't propagated out of the script run by `sys.remote_exec()`, so while `signal.raise_signal()` will work, if that signal handler sets an exception it'll just wind up as an unraisable exception and discarded.

---

Hm. This is sort of a PDB problem in general, isn't it?

Signal handlers only run on the main thread, so if you're using PDB in a multi-threaded program and you do
```py
while True:
    pass
```
on any thread but the main thread, that thread will be stuck forever, right?

The only thing that remote PDB is making worse is that getting stuck will happen for all N threads, instead of for N-1 of them 😄

---

> Exceptions aren't propagated out of the script run by `sys.remote_exec()`

It _might_ make sense to special case `KeyboardInterrupt` in particular and allow it to pass through. We have the eval loop swallow exceptions raised by the injected script because the main application isn't expecting them or prepared to handle them, but `KeyboardInterrupt` is special because it's always allowed to be raised from just about anywhere (at least if the app hasn't changed the default SIGINT handler), and it can't be raised accidentally by the injected script and would only be raised deliberately (_or_ if the user ctrl-c'd the injected script - but even in that case, propagating it is probably the best choice...)

@pablogsal mentioned to me earlier today that he has some ideas about better ways to do the interrupts, but we haven't had a chance to discuss it yet.

---

It's true for multi-thread programs. However, pdb is always horrible at debugging multi-thread programs. There are millions of issues there so that's not our major concern now. I really hope that a single-thread program can work with remote pdb.

---

Could this be fixed by running evaluated code with tracing enabled? Or is that impossible?

---

The issue is that the evaluated code is in a trace callback, so it won't call any other callback functions - otherwise it's infinite.

---

Right, but we could use `sys.call_tracing()` when calling that evaluated code - probably disabling the breakpoints first and reenabling them once it has finished executing. That should fix this issue, but I'm not sure what other issues it might lead to...

---

I think that will disable the breakpoints in the evaluated code?

---

Aren't they already disabled, since we're running the evaluated code from a trace function and not calling it with `sys.call_tracing()`?

If I do:
```py
(Pdb) import functools
(Pdb) b functools.cache
Breakpoint 1 at /opt/bb/lib/python3.13/functools.py:679
(Pdb) functools.cache(print)
<functools._lru_cache_wrapper object at 0x7f2309b5f1c0>
```
the breakpoint doesn't get hit.